frontend/src/pages/CompareV2.test.tsx

```diff
@@ -14,15 +14,266 @@
+      } as any,
+      match: {} as any,
+      toolbarProps: { actions: {}, breadcrumbs: [], pageTitle: '' },
+      updateBanner: updateBannerSpy,
+      updateDialog: () => null,
+      updateSnackbar: () => null,
+      updateToolbar: () => null,
+    };
+    return pageProps;
+  }
+
+  let runs: ApiRunDetail[] = [];
+
+  function newMockRun(id?: string): ApiRunDetail {
+    return {
+      pipeline_runtime: {
+        workflow_manifest: '{}',
+      },
+      run: {
+        id: id || 'test-run-id',
+        name: 'test run ' + id,
+        pipeline_spec: { pipeline_manifest: '' },
+      },
+    };
+  }
+
+  function newMockContext(name?: string, id?: number): Execution {
+    const context = new Context();
+    context.setName(name);
+    context.setId(id);
+    return context;
+  }
+
+  function newMockExecution(id?: number): Execution {
+    const execution = new Execution();
+    execution.setId(id);
+    execution
+      .getCustomPropertiesMap()
+      .set('display_name', new Value().setStringValue(`execution${id}`));
+    return execution;
+  }
+
+  function newMockEvent(id?: number): Execution {
+    const event = new Event();
+    event.setArtifactId(id);
+    event.setType(Event.Type.OUTPUT);
+    return event;
+  }
+
+  function newMockArtifact(id?: number): Execution {
+    const artifact = new Artifact();
+    artifact.setId(id);
+    return artifact;
+  }
+
+  it('getRun is called with query param IDs', async () => {
+    const getRunSpy = jest.spyOn(Apis.runServiceApi, 'getRun');
+    runs = [newMockRun(MOCK_RUN_1_ID), newMockRun(MOCK_RUN_2_ID), newMockRun(MOCK_RUN_3_ID)];
+    getRunSpy.mockImplementation((id: string) => runs.find(r => r.run!.id === id));
+
+    render(
+      <CommonTestWrapper>
+        <CompareV2 {...generateProps()} />
+      </CommonTestWrapper>,
+    );
+
+    expect(getRunSpy).toHaveBeenCalledWith(MOCK_RUN_1_ID);
+    expect(getRunSpy).toHaveBeenCalledWith(MOCK_RUN_2_ID);
+    expect(getRunSpy).toHaveBeenCalledWith(MOCK_RUN_3_ID);
+  });
+
+  it('Show page error on page when getRun request fails', async () => {
+    const getRunSpy = jest.spyOn(Apis.runServiceApi, 'getRun');
+    runs = [
+      newMockRun(MOCK_RUN_1_ID, true),
+      newMockRun(MOCK_RUN_2_ID, true),
+      newMockRun(MOCK_RUN_3_ID, true),
+    ];
+    getRunSpy.mockImplementation(_ => {
+      throw {
+        text: () => Promise.resolve('test error'),
+      };
+    });
+
+    render(
+      <CommonTestWrapper>
+        <CompareV2 {...generateProps()} />
+      </CommonTestWrapper>,
+    );
+    await TestUtils.flushPromises();
+
+    await waitFor(() =>
+      expect(updateBannerSpy).toHaveBeenLastCalledWith({
+        additionalInfo: 'test error',
+        message: 'Error: failed loading 3 runs. Click Details for more information.',
+        mode: 'error',
+      }),
+    );
+  });
+
+  it('Successful MLMD requests clear the banner', async () => {
```

Comment:
This is a very large test, but all of the `Spy` and `mockImplementation` code segments are necessary for the query to run through completely - should I remove some of the `expect` checks at the end to simplify this test?

Reply:
On further review, I realize that I can directly mock the `mlmdUtils` functions rather than the underlying requests, which would be better since I should be testing these functions independent of their implementation.